lib/spack/spack/detection/path.py

```diff
@@ -0,0 +1,143 @@
+# Copyright 2013-2021 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+"""PATH based detection of software installed in the system"""
+import collections
+import os
+import os.path
+import re
+
+import llnl.util.filesystem
+import llnl.util.tty
+
+import spack.util.environment
+
+from ._common import (
+    DetectedPackage,
+    _convert_to_iterable,
+    executable_prefix,
+    is_executable,
+)
```

Comment:
Why is `_common` private and this is not (since there is `detection.__init__`)?

Reply:
Ideally all call site should use the public API exposed in `spack.detection` and organizing the code in different modules is just for convenience of developers. I used a leading underscore for `_common` so that when reading `__init__.py` it's first in the list of relative imports since it contains code that can be used across different detection methods.

Up to now we have only detection based on paths, but we can add after this PR detection based on CrayPE modules and later detection based on `spack.json`.

Reply:
> used a leading underscore for _common so that when reading __init__.py it's first in the list of relative imports

I thought that the import order was determined by the line numbers: i.e. if you import `x` before `y` then the imports will be resolved in that order.

That being said, I think I understand the distinction that `detection.path` contains functions a user would want to run and `_common` does not (although in that case I'm not sure why those functions are made available in `__all__`).

Reply:
No, import order is resolved by, roughly:
1. set of things imported (standard library packages, before 3rd party libraries, before llnl.* before spack.*)
2. type of import (`import x [as y]` before `from x import y`)
3. lexicographically among the same set and type

Reply:
> Is this strictly required? I would assume that importing in any order would be OK

For anybody reading the conversation: this is not strictly required, but is just a way to have the subpackage with the common base classes / functions be at the top with the ordering imposed by `isort`. Renaming the module to `common.py` (without the underscore) as agreed on Slack with @scheibelp